ccl/backupccl: TestBackupWorkerFailure failed

ccl/backupccl.TestBackupWorkerFailure [failed](https://tanzanite.cluster.engflow.com/invocation/4554492b-d778-4517-93b5-6da66c531a9f) on master @ [2205334339583cea57db33e3c0d992f672e660e8](https://github.com/cockroachdb/cockroach/commits/2205334339583cea57db33e3c0d992f672e660e8):


```
=== RUN   TestBackupWorkerFailure
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestBackupWorkerFailure3861981193
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:124: automatically injected virtual cluster under test; see comment at top of test_server_shim.go for details.
    backup_test.go:8892: pq: could not reload backup manifest when retrying: reading backup checkpoint: recv msg error: grpc: open /var/lib/engflow/worker/work/2/exec/_tmp/d8ba48f76500d7ecd0a36e7fe2534e5e/TestBackupWorkerFailure870532821/worker-failure/2023/10/26-072028.82/progress/BACKUP-CHECKPOINT-13cdcfcdccd2cecfd2cdc9dfcfc8c5cdcfc5cdc6d1cfc6c6cfcdcdcccbc7dfd4cfcfcfcfdfaaabbcdf92c2d4cec9d1cdcecac9cec7cac7cdfffe864317050.tmp: no such file or directory [code 2/Unknown]
    testutils.go:289: no Invalid Descriptors
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestBackupWorkerFailure3861981193
--- FAIL: TestBackupWorkerFailure (7.61s)
```
<p>Parameters: <code>attempt=1</code>
, <code>run=9</code>
, <code>shard=6</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupWorkerFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-32757

ccl/backupccl.TestBackupWorkerFailure [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsGcpLinuxX8664BigVm_CclUnitTests/12738105?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsGcpLinuxX8664BigVm_CclUnitTests/12738105?buildTab=artifacts#/) on master @ [14733120c08ba1a0a050bb5442c3311090206566](https://github.com/cockroachdb/cockroach/commits/14733120c08ba1a0a050bb5442c3311090206566):


```
=== RUN   TestBackupWorkerFailure
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/6b441e4673fbb829e4ecf7b3ce3dc7db/logTestBackupWorkerFailure1838295603
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:159: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    backup_test.go:8893: pq: could not reload backup manifest when retrying: reading backup checkpoint: recv msg error: grpc: open /artifacts/tmp/_tmp/6b441e4673fbb829e4ecf7b3ce3dc7db/TestBackupWorkerFailure663766558/worker-failure/2023/11/18-072211.55/progress/BACKUP-CHECKPOINT-13cdcfcdccd2ceced2cec7dfcfc8c5cdcdc5ceced1c6cdcdc6cdcbccc7cadfd4cfcfcfcfdfaaabbcdf92c2d4ccc9d1cfc9cfcdcfc6cccdc8fffe2367661490.tmp: no such file or directory [code 2/Unknown]
    testutils.go:289: no Invalid Descriptors
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/6b441e4673fbb829e4ecf7b3ce3dc7db/logTestBackupWorkerFailure1838295603
--- FAIL: TestBackupWorkerFailure (12.58s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupWorkerFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/backupccl.TestBackupWorkerFailure [failed](https://tanzanite.cluster.engflow.com/invocation/8dd72aed-8fea-46dd-8ce5-44c52fa58893) on master @ [caff15394fcbe37208b46b2973714c27cc3a1417](https://github.com/cockroachdb/cockroach/commits/caff15394fcbe37208b46b2973714c27cc3a1417):


```
=== RUN   TestBackupWorkerFailure
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestBackupWorkerFailure14690034
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:159: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    backup_test.go:8895: pq: failed to run backup: exporting 7 ranges: exporting /Tenant/10/Table/106/1/{40-50}: failed to send RPC: sending to all replicas failed; last error: routing information detected to be stale: [NotLeaseHolderError] lease held by different store; r70: replica (n3,s3):2 not lease holder; current lease is repl=(n2,s2):3 seq=7 start=1700900250.678474222,0 epo=1 pro=1700900250.685620971,0
    testutils.go:289: no Invalid Descriptors
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestBackupWorkerFailure14690034
--- FAIL: TestBackupWorkerFailure (15.04s)
```
<p>Parameters: <code>attempt=1</code>
, <code>run=7</code>
, <code>shard=7</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupWorkerFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/backupccl.TestBackupWorkerFailure [failed](https://tanzanite.cluster.engflow.com/invocation/245b081b-ce22-4f63-8408-08a9e0d22d2a) on master @ [ba6423f916dc4006bb590c786cf53ad2b0213417](https://github.com/cockroachdb/cockroach/commits/ba6423f916dc4006bb590c786cf53ad2b0213417):


```
=== RUN   TestBackupWorkerFailure
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestBackupWorkerFailure711528825
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:159: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    backup_test.go:8830: pq: failed to run backup: running distributed backup to export 6 ranges: exporting /Tenant/10/Table/106/{1/90-2}: failed to send RPC: sending to all replicas failed; last error: routing information detected to be stale: [NotLeaseHolderError] lease held by different store; r88: replica (n3,s3):2 not lease holder; current lease is repl=(n1,s1):1 seq=5 start=1701164410.968016124,0 epo=1 pro=1701164410.970959034,0
    testutils.go:289: no Invalid Descriptors
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestBackupWorkerFailure711528825
--- FAIL: TestBackupWorkerFailure (9.34s)
```
<p>Parameters: <code>attempt=1</code>
, <code>run=10</code>
, <code>shard=6</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupWorkerFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/backupccl.TestBackupWorkerFailure [failed](https://tanzanite.cluster.engflow.com/invocation/3acdbe77-9948-41ba-99cc-a4b33c6edf54) on master @ [7f38906fd05002cecd9188bbeb1e9da4cb4811f4](https://github.com/cockroachdb/cockroach/commits/7f38906fd05002cecd9188bbeb1e9da4cb4811f4):


```
=== RUN   TestBackupWorkerFailure
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestBackupWorkerFailure723526812
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    backup_test.go:8857: pq: failed to run backup: running distributed backup to export 6 ranges: exporting /Tenant/10/Table/106/1/{60-70}: failed to send RPC: sending to all replicas failed; last error: routing information detected to be stale: [NotLeaseHolderError] lease held by different store; r84: replica (n3,s3):3 not lease holder; current lease is repl=(n1,s1):1 seq=5 start=1701844426.229274742,0 epo=1 pro=1701844426.255171001,0
    testutils.go:289: no Invalid Descriptors
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestBackupWorkerFailure723526812
--- FAIL: TestBackupWorkerFailure (14.22s)
```
<p>Parameters: <code>=</code>
, <code>attempt=1</code>
, <code>run=30</code>
, <code>shard=6</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #115486 ccl/backupccl: TestBackupWorkerFailure failed [C-test-failure O-robot P-1 T-disaster-recovery branch-release-23.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupWorkerFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/backupccl.TestBackupWorkerFailure [failed](https://tanzanite.cluster.engflow.com/invocation/928d4959-e0fd-4033-adc5-45796437ec62) on master @ [da982d571a4822164beeaa3c33e61f677c80a9fa](https://github.com/cockroachdb/cockroach/commits/da982d571a4822164beeaa3c33e61f677c80a9fa):


```
=== RUN   TestBackupWorkerFailure
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestBackupWorkerFailure2965708374
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    backup_test.go:8857: pq: failed to run backup: running distributed backup to export 7 ranges: exporting /Tenant/10/Table/106/1/{60-70}: failed to send RPC: sending to all replicas failed; last error: routing information detected to be stale: [NotLeaseHolderError] lease held by different store; r72: replica (n3,s3):3 not lease holder; current lease is repl=(n1,s1):1 seq=3 start=1701935128.681798308,0 exp=1701935134.681745228,0 pro=1701935128.681745228,0
    testutils.go:289: no Invalid Descriptors
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestBackupWorkerFailure2965708374
--- FAIL: TestBackupWorkerFailure (17.78s)
```
<p>Parameters: <code>=</code>
, <code>attempt=1</code>
, <code>run=13</code>
, <code>shard=6</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #115486 ccl/backupccl: TestBackupWorkerFailure failed [C-test-failure O-robot P-1 T-disaster-recovery branch-release-23.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupWorkerFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/backupccl.TestBackupWorkerFailure [failed](https://tanzanite.cluster.engflow.com/invocation/4cf4e7c1-c4d9-4d1b-a99a-a24f2e381015) on master @ [37ad01a3972cb4d34bfc6dfb4b9cfcac360b15dd](https://github.com/cockroachdb/cockroach/commits/37ad01a3972cb4d34bfc6dfb4b9cfcac360b15dd):


```
=== RUN   TestBackupWorkerFailure
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestBackupWorkerFailure1409636870
    test_log_scope.go:81: use -show-logs to present logs inline
    backup_test.go:8857: pq: failed to run backup: running distributed backup to export 6 ranges: deadline expired while waiting in queue: sql-sql-response, pri: normal-pri, deadline: 0001-01-01 00:00:00 +0000 UTC, start: 2023-12-08 06:35:29.20045302 +0000 UTC m=+15.057089379, dur: 4.954959ms
    testutils.go:289: no Invalid Descriptors
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestBackupWorkerFailure1409636870
--- FAIL: TestBackupWorkerFailure (9.07s)
```
<p>Parameters: <code>=</code>
, <code>attempt=1</code>
, <code>run=27</code>
, <code>shard=6</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #115486 ccl/backupccl: TestBackupWorkerFailure failed [C-test-failure O-robot P-1 T-disaster-recovery branch-release-23.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupWorkerFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/backupccl.TestBackupWorkerFailure [failed](https://tanzanite.cluster.engflow.com/invocation/63648dc0-5144-406e-b8ad-d6fd3ce448ab) on master @ [e747f6e6857a19d6048cb184b0f55c52cb8a6390](https://github.com/cockroachdb/cockroach/commits/e747f6e6857a19d6048cb184b0f55c52cb8a6390):


```
=== RUN   TestBackupWorkerFailure
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestBackupWorkerFailure2992059182
    test_log_scope.go:81: use -show-logs to present logs inline
    backup_test.go:8857: pq: failed to run backup: running distributed backup to export 8 ranges: exporting /Table/106/1/{20-30}: failed to send RPC: sending to all replicas failed; last error: routing information detected to be stale: [NotLeaseHolderError] lease held by different store; r74: replica (n3,s3):2 not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 exp=1702193660.154685819,0 pro=1702193654.154685819,0
    testutils.go:289: no Invalid Descriptors
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestBackupWorkerFailure2992059182
--- FAIL: TestBackupWorkerFailure (9.12s)
```

Parameters:
 - <code>=</code>
 - <code>attempt=1</code>
 - <code>run=15</code>
 - <code>shard=6</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #115486 ccl/backupccl: TestBackupWorkerFailure failed [C-test-failure O-robot P-1 T-disaster-recovery branch-release-23.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupWorkerFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/backupccl.TestBackupWorkerFailure [failed](https://tanzanite.cluster.engflow.com/invocation/6a1d80dd-378a-4121-a811-42b947a0e39f) on master @ [e747f6e6857a19d6048cb184b0f55c52cb8a6390](https://github.com/cockroachdb/cockroach/commits/e747f6e6857a19d6048cb184b0f55c52cb8a6390):


```
=== RUN   TestBackupWorkerFailure
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestBackupWorkerFailure4063365073
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    backup_test.go:8857: pq: failed to run backup: running distributed backup to export 8 ranges: exporting /Tenant/10/Table/106/1/{60-70}: failed to send RPC: sending to all replicas failed; last error: routing information detected to be stale: [NotLeaseHolderError] lease held by different store; r78: replica (n3,s3):2 not lease holder; current lease is repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1702276756.856726315,0
    testutils.go:289: no Invalid Descriptors
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestBackupWorkerFailure4063365073
--- FAIL: TestBackupWorkerFailure (8.37s)
```

Parameters:
 - <code>=</code>
 - <code>attempt=1</code>
 - <code>run=29</code>
 - <code>shard=6</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #115486 ccl/backupccl: TestBackupWorkerFailure failed [C-test-failure O-robot P-1 T-disaster-recovery branch-release-23.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupWorkerFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvclient/kvcoord/dist_sender.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender.go
@@ -1930,7 +1930,7 @@ func (ds *DistSender) sendPartialBatch(
 
 		if err != nil {
 			// Set pErr so that, if we don't perform any more retries, the
-			// deduceRetryEarlyExitError() call below the loop is inhibited.
+			// deduceRetryEarlyExitError() call below the loop includes this error.
 			pErr = kvpb.NewError(err)
 			switch {
 			case IsSendError(err):
@@ -2006,27 +2006,31 @@ func (ds *DistSender) sendPartialBatch(
 		break
 	}
 
-	// Propagate error if either the retry closer or context done
-	// channels were closed.
+	// Propagate error if either the retry closer or context done channels were
+	// closed. This replaces the return error since when the context is closed
+	// the underlying error is unpredictable and might have been retried.
+	if err := ds.deduceRetryEarlyExitError(ctx, pErr.GoError()); err != nil {
+		log.VErrEventf(ctx, 2, "replace error %s with %s", pErr, err)
+		pErr = kvpb.NewError(err)
+	}
+
 	if pErr == nil {
-		if err := ds.deduceRetryEarlyExitError(ctx); err == nil {
-			log.Fatal(ctx, "exited retry loop without an error")
-		} else {
-			pErr = kvpb.NewError(err)
-		}
+		log.Fatal(ctx, "exited retry loop without an error or early exit")
 	}
 
 	return response{pErr: pErr}
 }
 
-func (ds *DistSender) deduceRetryEarlyExitError(ctx context.Context) error {
+func (ds *DistSender) deduceRetryEarlyExitError(ctx context.Context, err error) error {
+	// We don't need to rewrap Ambiguous errors.
+	if errors.HasType(err, (*kvpb.AmbiguousResultError)(nil)) {
+		return nil
+	}
 	select {
 	case <-ds.rpcRetryOptions.Closer:
-		// Typically happens during shutdown.
-		return &kvpb.NodeUnavailableError{}
+		return errors.Wrapf(kvpb.NewAmbiguousResultError(errors.CombineErrors(&kvpb.NodeUnavailableError{}, err)), "aborted in DistSender")
 	case <-ctx.Done():
-		// Happens when the client request is canceled.
-		return errors.Wrap(ctx.Err(), "aborted in DistSender")
+		return errors.Wrapf(kvpb.NewAmbiguousResultError(errors.CombineErrors(ctx.Err(), err)), "aborted in DistSender")
 	default:
 	}
 	return nil

--- a/pkg/kv/kvclient/kvcoord/dist_sender_test.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender_test.go
@@ -623,6 +623,119 @@ func TestImmutableBatchArgs(t *testing.T) {
 	}
 }
 
+// TestErrorWithCancellationExit verifies that the DistSender never exits the
+// loop with a retriable error. These errors are not intended to escape the Send
+// and other code may not handle them correctly.
+func TestErrorWithCancellationExit(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	retriableErr := kvpb.NewError(
+		&kvpb.NotLeaseHolderError{
+			Replica: testUserRangeDescriptor.InternalReplicas[0],
+		})
+	terminalErr := kvpb.NewErrorf("boom")
+
+	tests := []struct {
+		name           string
+		retriableError *kvpb.Error
+		terminalError  *kvpb.Error
+		// Note that errorFn has side effects when it is called in the test.
+		cancelFn    func(ctx context.CancelFunc, stopper *stop.Stopper)
+		expectedErr string
+	}{
+		{
+			name:        "no error",
+			expectedErr: "",
+		},
+		{
+			name:           "terminal error",
+			retriableError: retriableErr,
+			terminalError:  terminalErr,
+			expectedErr:    "boom",
+		},
+		{
+			name:           "cancel context",
+			retriableError: retriableErr,
+			cancelFn: func(cancel context.CancelFunc, _ *stop.Stopper) {
+				// Cancel the context the request was started with.
+				cancel()
+			},
+			expectedErr: "context canceled",
+		},
+		{
+			name:           "stop stopper",
+			retriableError: retriableErr,
+			cancelFn: func(_ context.CancelFunc, stopper *stop.Stopper) {
+				// Stop the stopper simulating a shutdown.
+				stopper.Stop(context.Background())
+			},
+			expectedErr: "node unavailable",
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			retryCount := atomic.Int64{}
+			stopper := stop.NewStopper()
+			defer stopper.Stop(ctx)
+
+			clock := hlc.NewClockForTesting(nil)
+			rpcContext := rpc.NewInsecureTestingContext(ctx, clock, stopper)
+			g := makeGossip(t, stopper, rpcContext)
+			var testFn simpleSendFn = func(_ context.Context, _ *kvpb.BatchRequest) (*kvpb.BatchResponse, error) {
+				reply := &kvpb.BatchResponse{}
+
+				// Set a response so we don't get an out of bounds err in the non-error case.
+				var union kvpb.ResponseUnion
+				union.MustSetInner(&kvpb.PutResponse{})
+				reply.Responses = []kvpb.ResponseUnion{union}
+
+				// Count the number of times we are running.
+				count := retryCount.Add(1)
+
+				// Return a retriable error twice before running cancellation.
+				reply.Error = tc.retriableError
+				if count == 2 {
+					if tc.cancelFn != nil {
+						tc.cancelFn(cancel, stopper)
+					}
+				}
+
+				// Return retriable a few more times as cancellation may need to propagate.
+				if count > 5 {
+					reply.Error = tc.terminalError
+				}
+				return reply, nil
+			}
+
+			cfg := DistSenderConfig{
+				AmbientCtx: log.MakeTestingAmbientCtxWithNewTracer(),
+				Clock:      clock,
+				NodeDescs:  g,
+				Stopper:    stopper,
+				// Retry very quickly to make this test finish fast.
+				RPCRetryOptions: &retry.Options{
+					InitialBackoff: time.Millisecond,
+					MaxBackoff:     time.Millisecond,
+				},
+				TransportFactory:  adaptSimpleTransport(testFn),
+				RangeDescriptorDB: defaultMockRangeDescriptorDB,
+				Settings:          cluster.MakeTestingClusterSettings(),
+			}
+			ds := NewDistSender(cfg)
+			// Start a request that runs through distSender.
+			put := kvpb.NewPut(roachpb.Key("a"), roachpb.MakeValueFromString("value"))
+			_, pErr := kv.SendWrapped(ctx, ds, put)
+			if tc.expectedErr == "" {
+				require.Nil(t, pErr)
+			} else {
+				require.NotNil(t, pErr)
+				require.True(t, testutils.IsPError(pErr, tc.expectedErr))
+			}
+		})
+	}
+}
+
 // TestRetryOnNotLeaseHolderError verifies that the DistSender correctly updates
 // the leaseholder in the range cache and retries when receiving a
 // NotLeaseHolderError.

--- a/pkg/kv/kvclient/kvcoord/range_iter.go
+++ b/pkg/kv/kvclient/kvcoord/range_iter.go
@@ -230,7 +230,7 @@ func (ri *RangeIterator) Seek(ctx context.Context, key roachpb.RKey, scanDir Sca
 	}
 
 	// Check for an early exit from the retry loop.
-	if deducedErr := ri.ds.deduceRetryEarlyExitError(ctx); deducedErr != nil {
+	if deducedErr := ri.ds.deduceRetryEarlyExitError(ctx, err); deducedErr != nil {
 		ri.err = deducedErr
 	} else {
 		ri.err = errors.Wrapf(err, "RangeIterator failed to seek to %s", key)

--- a/pkg/kv/kvclient/kvcoord/txn_test.go
+++ b/pkg/kv/kvclient/kvcoord/txn_test.go
@@ -520,7 +520,7 @@ func TestTxnWriteReadConflict(t *testing.T) {
 		// Verify the expected blocking behavior.
 		if expBlocking {
 			require.Error(t, err)
-			require.ErrorIs(t, context.DeadlineExceeded, err)
+			require.ErrorIs(t, err, context.DeadlineExceeded)
 		} else {
 			require.NoError(t, err)
 			require.False(t, res.Exists())

--- a/pkg/kv/kvserver/client_merge_test.go
+++ b/pkg/kv/kvserver/client_merge_test.go
@@ -4187,7 +4187,7 @@ func TestStoreRangeMergeDuringShutdown(t *testing.T) {
 	// Send a dummy get request on the RHS to force a lease acquisition. We expect
 	// this to fail, as quiescing stores cannot acquire leases.
 	_, err = store.DB().Get(ctx, key.Next())
-	if exp := "not lease holder"; !testutils.IsError(err, exp) {
+	if exp := "node unavailable"; !testutils.IsError(err, exp) {
 		t.Fatalf("expected %q error, but got %v", exp, err)
 	}
 }

